File: backend/app/api/transactions.py
from datetime import date
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile
from sqlalchemy.orm import Session

from app.db.engine import get_db
from app.schemas.transaction import (
    ImportResult,
    TransactionCreate,
    TransactionListResponse,
    TransactionOut,
    TransactionUpdate,
)
from app.services.transaction import TransactionService

router = APIRouter()


@router.get("/transactions", response_model=TransactionListResponse)
def list_transactions(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: str | None = Query(None),
    sort_by: str = Query("date"),
    sort_dir: Literal["asc", "desc"] = Query("desc"),
    account_id: int | None = Query(None),
    category_id: int | None = Query(None),
    date_from: date | None = Query(None),
    date_to: date | None = Query(None),
    db: Session = Depends(get_db),
):
    service = TransactionService(db)
    return service.list_transactions(
        page=page,
        page_size=page_size,
        q=q,
        sort_by=sort_by,
        sort_dir=sort_dir,
        account_id=account_id,
        category_id=category_id,
        date_from=date_from,
        date_to=date_to,
    )


@router.post("/transactions", response_model=TransactionOut, status_code=201)
def create_transaction(data: TransactionCreate, db: Session = Depends(get_db)):
    service = TransactionService(db)
    return service.create_transaction(data)


@router.put("/transactions/{txn_id}", response_model=TransactionOut)
def update_transaction(txn_id: int, data: TransactionUpdate, db: Session = Depends(get_db)):
    service = TransactionService(db)
    result = service.update_transaction(txn_id, data)
    if not result:
        raise HTTPException(status_code=404, detail="Transaction not found")
    return result


@router.delete("/transactions/{txn_id}", status_code=204)
def delete_transaction(txn_id: int, db: Session = Depends(get_db)):
    service = TransactionService(db)
    deleted = service.delete_transaction(txn_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Transaction not found")


@router.post("/transactions/import", response_model=ImportResult)
async def import_transactions(file: UploadFile, db: Session = Depends(get_db)):
    content = await file.read()
    csv_text = content.decode("utf-8")

    # Strip optional header line like "Transactions For All Accounts..."
    lines = csv_text.split("\n")
    if lines and not lines[0].startswith("Date") and "Date" not in lines[0].split(",")[0]:
        csv_text = "\n".join(lines[1:])

    service = TransactionService(db)
    return service.import_csv(csv_text)

File: frontend/src/api/transactions.ts
import { api } from "./client";

export interface Transaction {
  id: number;
  date: string;
  description: string;
  raw_description: string | null;
  amount: number;
  category_id: number | null;
  account_id: number | null;
  is_recurring: boolean;
  tags: string | null;
  notes: string | null;
  created_at: string;
}

export interface TransactionListResponse {
  items: Transaction[];
  total: number;
  page: number;
  page_size: number;
}

export interface ImportResult {
  imported: number;
  skipped: number;
  accounts_created: number;
  categories_created: number;
}

export interface ListTransactionsParams {
  page?: number;
  page_size?: number;
  q?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
  account_id?: number;
  category_id?: number;
  date_from?: string;
  date_to?: string;
}

export interface CreateTransactionData {
  date: string;
  description: string;
  amount: number;
  category_id?: number | null;
  account_id?: number | null;
  is_recurring?: boolean;
  tags?: string | null;
  notes?: string | null;
}

export interface UpdateTransactionData {
  date?: string;
  description?: string;
  amount?: number;
  category_id?: number | null;
  account_id?: number | null;
  is_recurring?: boolean;
  tags?: string | null;
  notes?: string | null;
}

export function listTransactions(
  params?: ListTransactionsParams,
): Promise<TransactionListResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.page_size) search.set("page_size", String(params.page_size));
  if (params?.q) search.set("q", params.q);
  if (params?.sort_by) search.set("sort_by", params.sort_by);
  if (params?.sort_dir) search.set("sort_dir", params.sort_dir);
  if (params?.account_id) search.set("account_id", String(params.account_id));
  if (params?.category_id) search.set("category_id", String(params.category_id));
  if (params?.date_from) search.set("date_from", params.date_from);
  if (params?.date_to) search.set("date_to", params.date_to);
  const qs = search.toString();
  return api.get<TransactionListResponse>(`/api/transactions${qs ? `?${qs}` : ""}`);
}

export function createTransaction(data: CreateTransactionData): Promise<Transaction> {
  return api.post<Transaction>("/api/transactions", data);
}

export function updateTransaction(id: number, data: UpdateTransactionData): Promise<Transaction> {
  return api.put<Transaction>(`/api/transactions/${id}`, data);
}

export function deleteTransaction(id: number): Promise<void> {
  return api.delete<void>(`/api/transactions/${id}`);
}

export function importTransactions(file: File): Promise<ImportResult> {
  const formData = new FormData();
  formData.append("file", file);
  return api.postFile<ImportResult>("/api/transactions/import", formData);
}
